man/man1/nad2nad.1
.\" release 5
.\"
.ad b
.hy 1
.TH NAD2NAD 1 "2018/02/15 Rel. 5.0.0"
.SH NAME
nad2nad \- North American Datum conversion filter
.SH SYNOPSIS
.B nad2nad
[
.B \-eEfihortwW
[
.I args
] ] [
.B +args
]
file[s]
.SH DESCRIPTION
Program
.B nad2nad
is a filter to convert data between North
America Datum 1927 (NAD27) and North American Datum 1983.
.B nad2nad
can optionally process both State Plane Coordinate System (SPCS) and
Universal Transverse Mercator (UTM) grid data as well
as geographic data for both input and output.  This can also be 
accomplished with the \fIcs2cs\fR program.
.PP
The following control parameters can appear in any order:
.TP
.BI \-[i|o] " keyword[,keyword]"
The
.B \-i
and
.B \-o
option expect keyword arguments which define various characteristics and
processing modes of the
respective input data.
Usage allows multiple arguments to be included with a \- operator
when separated by commas.
Datum conversion requires the data to be in geographic coordinates, but
.B nad2nad
will allow conversion of data to and from SPCS or UTM grid systems.
The following are keywords and arguments recognized by both the
.B \-i
and
.B \-o
that will apply to respective input and output conversion of user
data to internal geographic coordinates:
.RS 1in
.TP
.B 27|83
datum of data
.TP
.BI utm= n
.B UTM
coordinates in meters for zone
.I n
.TP
.BI spcs= n
for data in
.B SPCS
coordinates, where
.I n
is state zone number.
.TP
.B feet
data units are in U.S. Surveyor's feet.
This is allowed only when the
.I spcs
option been previously used.
Default coordinates are in meters.
.TP
.B bin
for data in binary form.
.TP
.B rev
data in lat-lon order rather than default lon-lat order.
.TP
.BI hp= ss
use high precision conversion zone
.I ss.
Certain States have ancillary correction tables to further refine
the basic conus table.
.I Ss
key and States are:
.RS .5in
FL Florida
.br
MD Maryland
.br
TN Tennessee
.br
WI Wisconsin
.br
WO Washington, Oregon and northern part of California.
.RE
.RE
.TP
.BI \-t "a"
.I A
specifies a character employed as the first character to denote
a control line to be passed through without processing.
This option applicable to ascii input only.
(# is the default value).
.TP
.BI \-e " string"
.I String
is an arbitrary string to be output if an error is detected during
data transformations.
The default value is: *\et*.
Note that if the
.B "\-o bin"
option is employed, an error is output as HUGE_VAL for both values.
.TP
.BI \-r " region"
specifies which regional conversion table to employ which are identified
by the following:
.RS .5in
conus \- conterminous 48 States
.br
alaska \- State of Alaska
.br
hawaii \- State of Hawaii
.br
prvi \- Puerto Rico and Virgin Islands
.br
stgeorge \- St. George Is, Alaska
.br
stpaul \- St. Paul Is, Alaska
.br
stlrnc \- St. Lawrence Is, Alaska
.TP
.B \-E
Input coordinates are echoed to output before output values.
.RE
.TP
.BI \-f " format"
.I Format
is a
.I printf
format string to control the form of the output values.
For inverse projections, the output will be in degrees when this option
is employed.
If a format is specified for inverse projection the
output data will be in decimal degrees.
The default format is \(``%.2f\('' for forward projection and DMS
for inverse.
.TP
.BI \-[w|W] n
.I N
is the number of significant fractional digits to employ for
seconds output (when the option is not specified,
.B \-w3
is assumed).
When
.B \-W
is employed the fields will be constant width and with leading zeroes.
.PP
One or more
.I files
(processed in left to right order)
specify the source of data to be transformed.
A \- will specify the location of processing standard input.
If no files are specified, the input is assumed to be from
.I stdin.
For ASCII input data the two data values must be in the
first two white space separated fields and
when both input and output are ASCII all trailing portions
of the input line are appended to the output line.
.PP
Input geographic data
(longitude and latitude) must be in DMS format when neither
.I utm
nor
.I spcs
is specified, otherwise in meters or feet (\fIfeet\fR option used).
Input data fields must be separated by white space and not have
imbedded white space.
.PP
Output data will be in tab separated fields of DMS or grid
coordinates in meters or feet.
.PP
Any data after the two input values are echoed after the two
output data values.
.SH SEE ALSO
.B cs2cs(1U), proj(1U),
.br
.I "Cartographic Projection Procedures for the UNIX Environment\(emA User's Manual,"
(Evenden, 1990, Open-file report 90\-284).
.SH BUGS
A list of known bugs can found at https://github.com/OSGeo/proj.4/issues
where new bug reports can be submitted too.
.SH HOME PAGE
https://proj4.org/


NEWS
5.1.0 Release Notes
-------------------

 UPDATES
 -------

 o Function proj_errno_string() added to proj.h API (#847)

 o Validate units between pipeline steps and ensure transformation
   sanity (#906)

 o Print help when calling cct and gie without arguments (#907)

 o CITATION file added to source distribution (#914)

 o Webmercator operation added (#925)

 o Enhanced numerical precision of forward spherical Mercator near
   the Equator (#928)

 o Added --skip-lines option to cct (#923)

 o Consistently return NaN values on NaN input (#949)

 o Removed unused src/org_proj4_Projections.h file (#956)

 o Java Native Interface bindings updated (#957, #969)

 o Horizontal and vertical gridshift operations extended to
   the temporal domain (#1015)


 BUG FIXES
 ------------

 o Handle nan float cast overflow in PJ_robin.c and nad_intr.c (#887)

 o Avoid overflow when Horner order is unreasonably large (#893)

 o Avoid unwanted NaN conversions in etmerc (#899)

 o Avoid memory failure in gie when not specifying x,y,z in gie files (#902)

 o Avoid memory failure when +sweep is initialized incorrectly in geos (#908)

 o Return HUGE_VAL on erroneous input in ortho (#912)

 o Handle commented lines correctly in cct (#933)

 o Avoid segmentation fault when transformation coordinates outside grid
   area in deformation (#934)

 o Avoid doing false easting/northing adjustments on cartesian
   coordinates (#936)

 o Thread-safe creation of proj mutex (#954)

 o Avoid errors when setting up geos with +lat_0!=0 (#986)

 o Reset errno when running proj in verbose mode (#988)

 o Do not interpolate node values at nodata value in vertical
   grid shifts (#1004)

 o Restrict Horner degrees to positive integer values to avoid
   memory allocation issues (#1005)

 All bug fix numbers refer to issues or pull requests indexed at
 https://github.com/OSGeo/proj.4/

 THANKS TO
 ------------

 Version 5.1.0 is made possible by the following contributors:

   Kristian Evers
   Even Rouault
   Kurt Schwehr
   Mike Toews
   Martin Desruisseaux
   Charles Karney
   Thomas Knudsen
   Javier Goizueta
   Bas Couwenberg
   Adam Wulkiewicz
   Aaron Puchert

5.0.1 Release Notes
-------------------

 BUG FIXES
 ------------

 All bug fix numbers refer to issues or pull requests indexed at
 https://github.com/OSGeo/proj.4/

 o Handle ellipsoid change correctly in pipelines when
   +towgs84=0,0,0 is set #881

 o Handle the case where nad_ctable2_init returns NULL #883

 o Avoid shadowed declaration errors with old gcc #880

 o Expand +datum properly +datum in pipelines #872

 o Fail gracefully when incorrect headers are encountered in grid
   files #875

 o Improve roundtrip stability in pipelines using +towgs84 #871

 o Fixed typo in gie error codes #861

 o Numerical stability fixes to the geodesic package #826 #843

 o Make sure that transient errors are returned correctly #857

 o Make sure that locally installed header files are not used when
   building PROJ #849

 o Fix inconsistent parameter names in proj.h/proj_4D_api.c #842

 o Make sure +vunits is applied #833

 o Fix incorrect Web Mercator transformations #834

 THANKS TO
 ------------

 Version 5.0.1 is made possible by the following contributors:

   Mike Toews
   Kurt Schwehr
   Even Rouault
   Charles Karney
   Thomas Knudsen
   Kristian Evers


5.0.0 Release Notes
-------------------

This version of PROJ introduces some significant extensions and
improvements to (primarily) the geodetic functionality of the system.

The main driver for introducing the new features is the emergence of
dynamic reference frames, the increasing use of high accuracy GNSS,
and the related growing demand for accurate coordinate
transformations.  While older versions of PROJ included some geodetic
functionality, the new framework lays the foundation for turning PROJ
into a generic geospatial coordinate transformation engine.

The core of the library is still the well established projection code.
The new functionality is primarily exposed in a new programming
interface and a new command line utility, "cct" (for "Coordinate
Conversion and Transformation").  The old programming interface is
still available and can - to some extent - use the new geodetic
transformation features.

The internal architecture has also seen many changes and much
improvement.  So far, these improvements respect the existing
programming interface. But the process has revealed a need to simplify
and reduce the code base, in order to support sustained active
development.

!!!
!!! Therefore we have scheduled regular releases over the coming years
!!! which will gradually remove the old programming interface.
!!!
!!! This will cause breaking changes with the next two major version
!!! releases, which will affect all projects that depend on PROJ
!!! (cf. section "deprecations" below).
!!!

The decision to break the existing API has not been easy, but has
ultimately been deemed necessary to ensure the long term survival of
the project. Not only by improving the maintainability immensely, but
also by extending the potential user (and hence developer) community.

The end goal is to deliver a generic coordinate transformation
software package with a clean and concise code base appealing to
both users and developers.


VERSIONING AND NAMING
---------------------

For the first time in more than 25 years the major version number of
the software is changed. The decision to do this is based on the many
new features and new API. While backwards compatibility remains -
except in a few rare corner cases - the addition of a new and improved
programming interface warrants a new major release.

The new major version number unfortunately leaves the project in a bit
of a conundrum regarding the name. For the majority of the life-time
of the product it has been known as PROJ.4, but since we have now
reached version 5 the name is no longer aligned with the version
number.


src/nad2nad.c
/* <<<< North American Datum Transfer Program >>>> */
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#define PJ_LIST_H <nad_list.h>
#include <projects.h>
#include "emess.h"

#define MAX_LINE 1000
#define MAX_PARGS 100
#define PJ_INVERS(P) (P->inv ? 1 : 0)
	static int
echoin = 0,	/* echo input data to output line */
tag = '#';	/* beginning of line tag character */
	static char
*oform = (char *)0,	/* output format for x-y or decimal degrees */
*oterr = "*\t*",	/* output line for unprojectable input */
*inargs = 0,
*outargs = 0,
*czone = 0,
*usage =
"%s\nusage: %s [ -eEfihortwW [args] ] [ files ]\n";
	struct CTABLE
*ctab = 0,
*htab = 0;
static struct TAG_LIST {
	char *tag;
	short sw;
} ops_list[] = {
	"utm=",	0,
	"spcs=", 1,
	"feet", 2,
	"27", 3,
	"83", 4,
	"hp", 5,
	"bin", 6,
	"rev", 7,
	0, 0,
};
static struct IO_CON {
	short rev;	/* reverse lon/lat or x/y */
	short bin;	/* io binary */
	short ll;	/* io lat-lon */
	short t83;	/* data in 83 datum */
	short zone;	/* <100 utm zone, ==0 geog,  else state plane zone */
	short nprojc; /* number of entries in projc */
	char *hp;	/* high precision name */
	char *projc[10];	/* params for pj_init */
	PJ *cnv;
} input = {
	0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
	0,
}, output = {
	0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
	0,
};
	static void
set_zone(int in, struct IO_CON *io) {
	char tmp[20];

	if (io->hp) {
		io->t83 = 1;
		if (!(htab = nad_init(pj_get_default_ctx(), io->hp)))
			emess(1,"hp datum file: %s, failed: %s", io->hp,
				pj_strerrno(pj_errno));
	}
	if (io->zone > 0) {
		if (io->zone <= 60) { /* UTM zone */
			io->nprojc = 2; /* no other options allowed */
			io->projc[0] = "proj=utm";
			sprintf(tmp, "zone=%d", io->zone);
			io->projc[1] = io->t83 ? "ellps=GRS80" : "ellps=clrk66";
		} else /* SPCS zone */
			sprintf(tmp, "init=nad%s:%d", io->t83 ? "83" : "27", io->zone);
		io->projc[io->nprojc++] = tmp;
		io->projc[io->nprojc++] = "no_defs";
		if (!(io->cnv = pj_init(io->nprojc, io->projc)))
			emess(1,pj_strerrno(pj_errno));
		io->ll = 0;
	}
}
	static void
setup() {
	/* check and set zone operations */
	if (input.hp && output.hp)
		emess(1,"both input and output cannot be high precision");
	set_zone(1, &input);
	set_zone(0, &output);
	if (input.cnv && !output.cnv)
		output.ll = 1;
	if (output.cnv && !input.cnv)
		input.ll = 1;
	if (!input.cnv && !output.cnv)
		output.ll = input.ll = 1;
	if (czone) {
		if (!input.hp && !output.hp && input.t83 == output.t83)
			emess(1,"identical datums");
		if (!(ctab = nad_init(pj_get_default_ctx(), czone)))
			emess(1,"datum file: %s, failed: %s", czone, pj_strerrno(pj_errno));
	} else if (input.t83 != output.t83)
		emess(1,"conversion region (-r) not specified");
}
	static void
set_ops(char *s, struct IO_CON *io) {
	char *intag;
	struct TAG_LIST *p;

	for ( ; (intag = strtok(s, " ,\t")) != NULL; s = 0) {
		for (p = ops_list; p->tag; ++p) {
			if (!strncmp(intag, p->tag, strlen(p->tag)))
				break;
		}
		if (!p->tag)
			emess(1,"invalid selection");
		switch (p->sw) {
		case 0:
		case 1:
			s = strchr(intag, '=') + 1;
			io->zone = atoi(s);
			break;
		case 2:
			if (io->zone <= 60)
				emess(1,"spcs zone must be selected");
			io->projc[io->nprojc++] = "units=us-ft";
			break;
		case 3: io->t83 = 0; break;
		case 4: io->t83 = 1; break;
		case 5:
			if (!(intag = strchr(intag, '=')) || *++intag == '\0')
				emess(1,"hp missing name");
			strcpy(io->hp = (char*)malloc(strlen(intag)+1), intag);
			break;
		case 6: io->bin = 1; break;
		case 7: io->rev = 1; break;
		}
	}
}
	static void
process(FILE *fid) {
	char line[MAX_LINE], *s, t, pline[100];
	projUV val;
	double tmp;

	for (;;) {
		if (input.bin)
			fread(&val, sizeof(projUV), 1, fid);
		else if ((s = fgets(line, MAX_LINE, fid)) != NULL) {
			if (*s == tag) {
				fputs(line, stdout);
				continue;
			} else if (input.ll) {
				val.u = dmstor(s, &s);
				val.v = dmstor(s, &s);
			} else {
				val.u = strtod(s, &s);
				val.v = strtod(s, &s);
			}
		}
		if (feof(fid))
			break;
		if (input.rev) {
			tmp = val.u;
			val.u = val.v;
			val.v = tmp;
		}
		/* data in, manupulate */
		if (input.cnv)
			val = pj_inv(val, input.cnv);
		if (input.hp)
			val = nad_cvt(val, 1, htab);
		/* nad conversion */
		if (ctab)
			val = nad_cvt(val, input.t83 ? 1 : 0, ctab);
		if (output.hp)
			val = nad_cvt(val, 0, htab);
		if (output.cnv)
			val = pj_fwd(val, output.cnv);
		/* output data */
		if (output.rev) {
			tmp = val.u;
			val.u = val.v;
			val.v = tmp;
		}
		if (output.bin)
			(void)fwrite(&val, sizeof(projUV), 1, stdout);
		else {
			if (echoin) {
				t = *s;
				*s = '\0';
				(void)fputs(line, stdout);
				(void)putchar('\t');
				*s = t;
			}
			if (val.u == HUGE_VAL)
				(void)fputs(oterr, stdout);
			else if (output.ll)
				if (oform) {
					(void)printf(oform, val.u * RAD_TO_DEG);
					(void)putchar('\t');
					(void)printf(oform, val.v * RAD_TO_DEG);
				} else if (output.rev) {
					(void)fputs(rtodms(pline, val.u, 'N', 'S'), stdout);
					(void)putchar('\t');
					(void)fputs(rtodms(pline, val.v, 'E', 'W'), stdout);
				} else {
					(void)fputs(rtodms(pline, val.u, 'E', 'W'), stdout);
					(void)putchar('\t');
					(void)fputs(rtodms(pline, val.v, 'N', 'S'), stdout);
				}
			else {
				(void)printf(oform ? oform : "%.2f", val.u);
				(void)putchar('\t');
				(void)printf(oform ? oform : "%.2f", val.v);
			}
			if (input.bin)
				putchar('\n');
			else
				(void)fputs(s, stdout);
		}
	}
}

int main(int argc, char **argv) {
	char *arg, **eargv = argv, work[MAX_PARGS];
	FILE *fid;
	int eargc = 0, c;

	if ((emess_dat.Prog_name = strrchr(*argv,DIR_CHAR)) != NULL)
		++emess_dat.Prog_name;
	else emess_dat.Prog_name = *argv;
	if (argc <= 1 ) {
		(void)fprintf(stderr, usage, pj_release, emess_dat.Prog_name);
		exit (0);
	}
		/* process run line arguments */
	while (--argc > 0) { /* collect run line arguments */
		if(**++argv == '-') for(arg = *argv;;) {
			switch(*++arg) {
			case '\0': /* position of "stdin" */
				if (arg[-1] == '-') eargv[eargc++] = "-";
				break;
			case 'i': /* input control */
			case 'o': /* output control */
				if (--argc <= 0) goto noargument;
				strncpy(work, *++argv, MAX_PARGS);
				set_ops(work, *arg == 'i' ? &input : &output);
				continue;
			case 'r': /* nad27/83 conversion zone */
				if (--argc <= 0) goto noargument;
				czone = *++argv;
				continue;
			case 'E': /* echo ascii input to ascii output */
				echoin = 1;
				continue;
			case 't': /* set col. one char */
				if (arg[1]) tag = *++arg;
				else emess(1,"missing -t col. 1 tag");
				continue;
			case 'W': /* specify seconds precision */
			case 'w': /* -W for constant field width */
				if ((c = arg[1]) != 0 && isdigit(c)) {
					set_rtodms(c - '0', *arg == 'W');
					++arg;
				} else
				    emess(1,"-W argument missing or non-digit");
				continue;
			case 'f': /* alternate output format degrees or xy */
				if (--argc <= 0) goto noargument;
				oform = *++argv;
				continue;
			case 'e': /* error line alternative */
				if (--argc <= 0)
noargument:			   emess(1,"missing argument for -%c",*arg);
				oterr = *++argv;
				continue;
			default:
				emess(1, "invalid option: -%c",*arg);
				break;
			}
			break;
		} else /* assumed to be input file name(s) */
			eargv[eargc++] = *argv;
	}
	if (eargc == 0) /* if no specific files force sysin */
		eargv[eargc++] = "-";
	/* done with parameter and control input */
	setup();
	/* process input file list */
	for ( ; eargc-- ; ++eargv) {
		if (**eargv == '-') {
			fid = stdin;
			emess_dat.File_name = "<stdin>";
		} else {
			if ((fid = fopen(*eargv, "r")) == NULL) {
				emess(-2, *eargv, "input file");
				continue;
			}
			emess_dat.File_name = *eargv;
		}
		emess_dat.File_line = 0;
		/* process file */
		process(fid);
		(void)fclose(fid);
		emess_dat.File_name = 0;
	}
	exit(0); /* normal completion */
}
